File: 程式設計：資料結構答案檔/Go/301.go
package main

import (
	"encoding/csv"
	"fmt"
	"os"
	"sort"
	"strconv"
)

type Player struct {
	Id    int
	Name  string
	Level int
	Score int
}

func readCsv(filePath string) ([]Player, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	csvReader := csv.NewReader(file)
	records, err := csvReader.ReadAll()
	if err != nil {
		return nil, err
	}

	var players []Player
	for _, record := range records[1:] {
		id, _ := strconv.Atoi(record[0])
		level, _ := strconv.Atoi(record[2])
		score, _ := strconv.Atoi(record[3])
		players = append(players, Player{Id: id, Name: record[1], Level: level, Score: score})
	}

	return players, nil
}

func main() {
	filePath := "read.csv"
	players, err := readCsv(filePath)
	if err != nil {
		fmt.Println("Error reading CSV file:", err)
		return
	}

	var n int
	_, err = fmt.Scanf("%d", &n)
	if err != nil {
		fmt.Println("Error reading input:", err)
		return
	}

	sort.Slice(players, func(i, j int) bool {
		if players[i].Score == players[j].Score {
			return players[i].Id > players[j].Id
		}
		return players[i].Score > players[j].Score
	})

	if n > len(players) {
		fmt.Println("Exceeds data upper limit")
	} else {
		for i := 0; i < n; i++ {
			fmt.Printf("%d | %d | %s | %d | %d\n", i+1, players[i].Id, players[i].Name, players[i].Level, players[i].Score)
		}
	}
}

File: 程式設計：資料結構答案檔/C/301.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_PLAYERS 1000
#define MAX_NAME_LENGTH 50

typedef struct {
    int Id;
    char Name[MAX_NAME_LENGTH];
    int Level;
    int Score;
} Player;

int comparePlayers(const void *a, const void *b) {
    Player *player1 = (Player *)a;
    Player *player2 = (Player *)b;

    return player2->Score - player1->Score;
}

void readCsvFile(const char* filePath, Player players[], int* count) {
    FILE* file = fopen(filePath, "r");
    if (!file) {
        printf("Error opening file\n");
        return;
    }

    char line[256];
    fgets(line, sizeof(line), file);

    while (fgets(line, sizeof(line), file) && *count < MAX_PLAYERS) {
        Player p;
        if (sscanf(line, "%d,%49[^,],%d,%d", &p.Id, p.Name, &p.Level, &p.Score) == 4) {
            players[(*count)++] = p;
        }
    }

    fclose(file);
}

void displaySortedPlayerData(Player players[], int n) {
    for (int i = 0; i < n; i++) {
        printf("%d | %d | %s | %d | %d\n", i + 1, players[i].Id, players[i].Name, players[i].Level, players[i].Score);
    }
}

int main() {
    Player players[MAX_PLAYERS];
    int count = 0;
    readCsvFile("read.csv", players, &count);

    int n;
    scanf("%d", &n);

    if (n > count) {
        printf("Exceeds data upper limit\n");
    } else {
        qsort(players, count, sizeof(Player), comparePlayers);
        displaySortedPlayerData(players, n);
    }

    return 0;
}
